File: day9/python/main.py
from enum import Enum
from queue import Queue, Empty
import itertools
import operator

import numpy as np


def input_(stdin):
    return stdin


def output_(input_value):
    return input_value


def jump_if_true(condition, target):
    if condition != 0:
        return target
    else:
        return None


def jump_if_false(condition, target):
    if condition == 0:
        return target
    else:
        return None


def less_than(x, y):
    return int(x < y)


def equals(x, y):
    return int(x == y)


def relative_base_offset(val):
    return val


def noop():
    pass


class Output(Enum):
    VAL = 0
    PTR = 1
    NONE = 2
    STDOUT = 3
    BASE = 4
    HALT = 5


# op_code: (input_param_count, takes_stdin, output_type, func)
OP_CODES = {
    1: (2, False, Output.VAL, operator.add),
    2: (2, False, Output.VAL, operator.mul),
    3: (0, True, Output.VAL, input_),
    4: (1, False, Output.STDOUT, output_),
    5: (2, False, Output.PTR, jump_if_true),
    6: (2, False, Output.PTR, jump_if_false),
    7: (2, False, Output.VAL, less_than),
    8: (2, False, Output.VAL, equals),
    9: (1, False, Output.BASE, relative_base_offset),
    99: (0, False, Output.HALT, noop),
}


class Memory(dict):
    def __init__(self, list_data):
        for i, data in enumerate(list_data):
            self[i] = data

    def __getitem__(self, key):
        if isinstance(key, slice):
            range_ = range(key.start or 0, key.stop or len(self), key.step or 1)
            return [self[k] for k in range_]
        else:
            return super().__getitem__(key)

    def __missing__(self, key):
        if key < 0:
            raise IndexError(f"Invalid memory address: {key}")
        return 0


class Program:
    def __init__(self, memory_data, stdin, stdout):
        self._memory = Memory(memory_data)
        self._instruction_ptr = 0
        self._relative_base = 0
        self._stdin = stdin
        self._stdout = stdout
        self._halted = False

    def halted(self):
        return self._halted

    def _get_input_params(self, input_param_count, param_modes):
        if not input_param_count:
            return []

        input_ptr = self._instruction_ptr + 1
        input_memory = self._memory[input_ptr:input_ptr + input_param_count]
        param_modes = list(reversed([int(mode) for mode in str(param_modes)]))

        res = []
        for i, param in enumerate(input_memory):
            if i < len(param_modes):
                mode = param_modes[i]
            else:
                mode = 0
            if mode == 0:
                # Position mode: Interpret param as a memory address
                if param < 0:
                    raise RuntimeError(f"Invalid memory address: {param}")
                res.append(self._memory[param])
            elif mode == 1:
                # Immediate mode: Interpret param as a value
                res.append(param)
            elif mode == 2:
                # Relative mode: Interpret param as a memory address relative to base
                if self._relative_base + param < 0:
                    raise RuntimeError(
                        f"Invalid memory address: {self._relative_base + param}"
                    )
                res.append(self._memory[self._relative_base + param])
            else:
                raise ValueError(f"Invalid parameter mode: mode")
        return res

    def run(self):
        while True:
            # Read op_code, parameters and stdin
            param_modes, op_code = divmod(self._memory[self._instruction_ptr], 100)
            input_param_count, takes_stdin, output_type, func = OP_CODES[op_code]
            input_values = self._get_input_params(input_param_count, param_modes)
            if takes_stdin:
                try:
                    input_values.append(self._stdin.get_nowait())
                except Empty:
                    break

            # Run operation and handle its output
            output = func(*input_values)
            step_size = 1 + input_param_count
            if output_type == Output.NONE:
                pass
            elif output_type == Output.VAL:
                output_param_mode = (
                    self._memory[self._instruction_ptr] // 10**(input_param_count + 2)
                )
                output_addr = self._memory[self._instruction_ptr + input_param_count + 1]
                if output_param_mode == 2:
                    output_addr += self._relative_base
                self._memory[output_addr] = output
                step_size = 2 + input_param_count
            elif output_type == Output.PTR:
                if output is None:
                    pass
                else:
                    self._instruction_ptr = output
                    step_size = None
            elif output_type == Output.STDOUT:
                self._stdout.put_nowait(output)
            elif output_type == Output.BASE:
                self._relative_base += output
            elif output_type == Output.HALT:
                self._halted = True
                break

            # Step instruction pointer
            if step_size is not None:
                self._instruction_ptr += step_size


input_file = "../input.txt"
with open(input_file, 'r') as f:
    input_text = f.read()

# Test:
# input_text = "109,1,204,-1,1001,100,1,100,1008,100,16,101,1006,101,0,99"
# input_text = "1102,34915192,34915192,7,4,7,99,0"
# input_text = "104,1125899906842624,99"
# input_text = "109,2,203,-1,204,-1,99"
# input_text = "109,-1,4,1,99"
memory = np.array([int(i) for i in input_text.split(',')])

stdin = Queue()
stdout = Queue()
program = Program(memory, stdin, stdout)
stdin.put(1)
program.run()
assert program.halted()
output = stdout.get()
assert stdout.empty()

print("BOOST keycode (part 1):", output)


program = Program(memory, stdin, stdout)
stdin.put(2)
program.run()
assert program.halted()
output = stdout.get()
assert stdout.empty()
print("Distress beacon coordinates (part 2):", output)

File: day9/rust/src/main.rs
use num::{self, Integer};
use num_derive::FromPrimitive;
use num_traits::FromPrimitive;
use std::fs;
use std::sync::mpsc::{channel, Receiver, Sender};
use std::thread;

type Value = i128;

/// Split number at `pos` in decimal representation.
fn split_num<T>(n: T, pos: usize) -> (T, T)
where
    T: Integer + Clone + Copy + From<i32>,
{
    let divisor = num::pow(10i32, pos).into();
    let div = n / divisor;
    let rem = n % divisor;
    (div, rem)
}

/// Split all param modes
fn split_param_modes(param_modes: Value, input_count: Value) -> (Vec<Value>, Value) {
    let mut rest = param_modes;
    let mut input_modes = Vec::with_capacity(input_count as usize);
    for _ in 0..input_count {
        let (r, mode) = split_num(rest, 1);
        input_modes.push(mode);
        rest = r;
    }
    assert!(rest <= 2);
    (input_modes, rest)
}

#[derive(FromPrimitive)]
enum OpCode {
    Add = 1,
    Mul = 2,
    Input = 3,
    Output = 4,
    JumpIfTrue = 5,
    JumpIfFalse = 6,
    LessThan = 7,
    Equals = 8,
    RelativeBaseOffset = 9,
    Halt = 99,
}

/// An implementation of the IntCode program interface from advent of code 2019.
///
/// Stdin and stdout are handled via channels.
pub struct Program {
    instruction_ptr: usize,
    relative_base: Value,
    memory: Vec<Value>,
    stdin: Receiver<Value>,
    stdout: Sender<Value>,
}

impl Program {
    pub fn new(memory: Vec<Value>, stdin: Receiver<Value>, stdout: Sender<Value>) -> Program {
        Program {
            instruction_ptr: 0,
            relative_base: 0,
            memory: memory,
            stdin: stdin,
            stdout: stdout,
        }
    }

    /// Return the Value at current instruction_ptr address while also advancing
    /// the instruction_ptr by one.
    fn next_mem(&mut self) -> Value {
        let val = self.read_mem(self.instruction_ptr);
        self.instruction_ptr += 1;
        val
    }

    /// Grow the memory to include addr and initialize all new memory addresses
    /// to zero.
    fn init_mem(&mut self, addr: usize) {
        self.memory.resize(addr + 1, 0);
    }

    /// Read from memory address `addr`, growing the memory if needed.
    /// Previously uninitialized memory is initialized to zero.
    fn read_mem(&mut self, addr: usize) -> Value {
        if addr >= self.memory.len() {
            self.init_mem(addr);
        }
        self.memory[addr]
    }

    /// Write `value` to memory address `addr`, growing the memory if needed.
    fn write_mem(&mut self, addr: usize, value: Value) {
        if addr >= self.memory.len() {
            self.init_mem(addr);
        }
        self.memory[addr] = value;
    }

    /// Read a value from memory starting at the instruction pointer for each
    /// parameter mode.
    ///
    /// Parameter modes are interpreted as follows:
    ///   * 0 (Address mode) : Interpret the value as an address and return the
    ///                        value from that address.
    ///   * 1 (Immediate mode) : Return the value directly.
    ///   * 2 (Relative mode) : Interpret the value as an address relative to
    ///                         current relative base and return the value from
    ///                         that address.
    fn read(&mut self, param_modes: Vec<Value>) -> Vec<Value> {
        let mut result = Vec::with_capacity(param_modes.len());
        for mode in param_modes {
            let val = self.next_mem();
            result.push(match mode {
                0 => {
                    assert!(val >= 0);
                    self.read_mem(val as usize)
                }
                1 => val,
                2 => {
                    assert!(self.relative_base + val >= 0);
                    self.read_mem((self.relative_base + val) as usize)
                }
                _ => panic!("Invalid input parameter mode: {}", mode),
            });
        }
        result
    }

    /// Write `val` to the address at instruction pointer and advance the
    /// instruction pointer by one.
    ///
    /// Parameter modes are interpreted as follows:
    ///   * 0 (Address mode) : Interpret the value as an address and write to
    ///                         that address.
    ///   * 2 (Relative mode) : Interpret the value as an address relative to
    ///                         current relative base and write to that address.
    fn write(&mut self, val: Value, param_mode: Value) {
        let addr = self.next_mem()
            + match param_mode {
                0 => 0,
                2 => self.relative_base,
                _ => panic!("Invalid output parameter mode: {}", param_mode),
            };
        assert!(addr >= 0);
        self.write_mem(addr as usize, val);
    }

    /// Start running the program, returning after reaching a halt instruction.
    pub fn run(&mut self) {
        loop {
            let (param_modes, op_code) = split_num(self.next_mem(), 2);
            match FromPrimitive::from_i128(op_code) {
                Some(OpCode::Add) => {
                    let (input_param_modes, output_param_mode) = split_param_modes(param_modes, 2);
                    let sum: Value = self.read(input_param_modes).iter().sum();
                    self.write(sum, output_param_mode);
                }
                Some(OpCode::Mul) => {
                    let (input_param_modes, output_param_mode) = split_param_modes(param_modes, 2);
                    let product: Value = self.read(input_param_modes).iter().product();
                    self.write(product, output_param_mode);
                }
                Some(OpCode::Input) => {
                    let (_, output_param_mode) = split_param_modes(param_modes, 0);
                    let val: Value = self.stdin.recv().unwrap();
                    self.write(val, output_param_mode);
                }
                Some(OpCode::Output) => {
                    let (input_param_modes, _) = split_param_modes(param_modes, 1);
                    let val = self.read(input_param_modes).pop().unwrap();
                    self.stdout.send(val).unwrap();
                }
                Some(OpCode::JumpIfTrue) => {
                    let (input_param_modes, _) = split_param_modes(param_modes, 2);
                    let mut params = self.read(input_param_modes);
                    let new_addr = params.pop().unwrap();
                    let condition = params.pop().unwrap();
                    if condition != 0 {
                        assert!(new_addr >= 0);
                        self.instruction_ptr = new_addr as usize;
                    }
                }
                Some(OpCode::JumpIfFalse) => {
                    let (input_param_modes, _) = split_param_modes(param_modes, 2);
                    let mut params = self.read(input_param_modes);
                    let new_addr = params.pop().unwrap();
                    let condition = params.pop().unwrap();
                    if condition == 0 {
                        assert!(new_addr >= 0);
                        self.instruction_ptr = new_addr as usize;
                    }
                }
                Some(OpCode::LessThan) => {
                    let (input_param_modes, output_param_mode) = split_param_modes(param_modes, 2);
                    let mut params = self.read(input_param_modes);
                    let param2 = params.pop().unwrap();
                    let param1 = params.pop().unwrap();
                    let val = (param1 < param2) as Value;
                    self.write(val, output_param_mode);
                }
                Some(OpCode::Equals) => {
                    let (input_param_modes, output_param_mode) = split_param_modes(param_modes, 2);
                    let mut params = self.read(input_param_modes);
                    let param2 = params.pop().unwrap();
                    let param1 = params.pop().unwrap();
                    let val = (param1 == param2) as Value;
                    self.write(val, output_param_mode);
                }
                Some(OpCode::RelativeBaseOffset) => {
                    let (input_param_modes, _) = split_param_modes(param_modes, 1);
                    let mut params = self.read(input_param_modes);
                    let offset = params.pop().unwrap();
                    self.relative_base += offset;
                }
                Some(OpCode::Halt) => break,
                _ => panic!("Unimplemented OpCode: {}", op_code),
            }
        }
    }
}

fn read_input_file() -> Vec<Value> {
    let input_file = "../input.txt";
    fs::read_to_string(input_file)
        .unwrap()
        .split(",")
        .map(|s| s.trim().parse().unwrap())
        .collect()
}

fn main() {
    let memory = read_input_file();
    let (stdin_sender, stdin_receiver) = channel();
    let (stdout_sender, stdout_receiver) = channel();
    let mut program = Program::new(memory, stdin_receiver, stdout_sender);
    // stdin_sender.send(1).unwrap();
    stdin_sender.send(2).unwrap();
    thread::spawn(move || program.run());

    for output in stdout_receiver {
        println!("{}", output)
    }

    // println!("Total cycle length (part 2): {}", total_cycle);
}

#[cfg(test)]
mod tests {
    use super::*;

    fn create_program(mem: Vec<Value>) -> Program {
        let (_stdin_sender, stdin_receiver) = channel();
        let (stdout_sender, _stdout_receiver) = channel();
        Program::new(mem, stdin_receiver, stdout_sender)
    }

    #[test]
    fn test_split_num() {
        assert_eq!(split_num(1, 0), (1, 0));
        assert_eq!(split_num(12, 0), (12, 0));
        assert_eq!(split_num(12, 1), (1, 2));
        assert_eq!(split_num(12, 2), (0, 12));
        assert_eq!(split_num(12, 3), (0, 12));
    }

    #[test]
    fn test_split_param_modes() {
        assert_eq!(split_param_modes(0, 0), (vec![], 0));
        assert_eq!(split_param_modes(0, 1), (vec![0], 0));
        assert_eq!(split_param_modes(0, 3), (vec![0, 0, 0], 0));
        assert_eq!(split_param_modes(210, 2), (vec![0, 1], 2));
    }

    #[test]
    fn test_read_mem_mode0() {
        let mut program = create_program(vec![-1, 0]);
        program.instruction_ptr = 1;
        assert_eq!(program.read(vec![0]), vec![-1]);
    }

    #[test]
    fn test_read_mem_mode1() {
        let mut program = create_program(vec![-1, 0]);
        program.instruction_ptr = 1;
        assert_eq!(program.read(vec![1]), vec![0]);
    }

    #[test]
    fn test_read_mem_mode2() {
        let mut program = create_program(vec![-1, 0, 1]);
        program.instruction_ptr = 1;
        program.relative_base = 2;
        assert_eq!(program.read(vec![2]), vec![1]);
    }

    #[test]
    fn test_read_mem_multiple() {
        let mut program = create_program(vec![-1, 1, 2, 3, 5]);
        program.instruction_ptr = 1;
        program.relative_base = -1;
        assert_eq!(program.read(vec![2, 1, 0]), vec![-1, 2, 3]);
    }

    #[test]
    fn test_read_outside_mem() {
        let mut program = create_program(vec![-1, 2]);
        program.instruction_ptr = 3;
        program.read(vec![0]);
        assert_eq!(program.memory, vec![-1, 2, 0, 0]);
    }

    #[test]
    fn test_write_mem_mode0() {
        let mut program = create_program(vec![-1, 0]);
        program.instruction_ptr = 1;
        program.write(1, 0);
        assert_eq!(program.memory, vec![1, 0]);
    }

    #[test]
    fn test_write_mem_mode2() {
        let mut program = create_program(vec![-1, 0, 0]);
        program.instruction_ptr = 1;
        program.relative_base = 1;
        program.write(1, 2);
        assert_eq!(program.memory, vec![-1, 1, 0]);
    }

    #[test]
    fn test_write_outside_mem() {
        let mut program = create_program(vec![-1, 2]);
        program.instruction_ptr = 3;
        program.write(1, 0);
        assert_eq!(program.memory, vec![1, 2, 0, 0]);
    }

    #[test]
    fn test_op1_add_mode0() {
        let mut program = create_program(vec![1, 0, 4, 5, 99, -1]);
        program.run();
        assert_eq!(program.memory, vec![1, 0, 4, 5, 99, 100]);
    }

    #[test]
    fn test_op1_add_mode1() {
        let mut program = create_program(vec![1101, 50, 50, 5, 99, -1]);
        program.run();
        assert_eq!(program.memory, vec![1101, 50, 50, 5, 99, 100]);
    }

    #[test]
    fn test_op1_add_mode2() {
        let mut program = create_program(vec![22201, 2, 3, 3, 99, -1]);
        program.relative_base = 2;
        program.run();
        assert_eq!(program.memory, vec![22201, 2, 3, 3, 99, 98]);
    }

    #[test]
    fn test_op2_mul() {
        let mut program = create_program(vec![1102, 5, 3, 5, 99, -1]);
        program.run();
        assert_eq!(program.memory, vec![1102, 5, 3, 5, 99, 15]);
    }

    #[test]
    fn test_op3_input() {
        let (stdin_sender, stdin_receiver) = channel();
        let (stdout_sender, _stdout_receiver) = channel();
        let mut program = Program::new(vec![3, 3, 99, -1], stdin_receiver, stdout_sender);
        stdin_sender.send(5).unwrap();
        program.run();
        assert_eq!(program.memory, vec![3, 3, 99, 5]);
    }

    #[test]
    fn test_op4_output() {
        let (_stdin_sender, stdin_receiver) = channel();
        let (stdout_sender, stdout_receiver) = channel();
        let mut program = Program::new(vec![4, 2, 99], stdin_receiver, stdout_sender);
        program.run();
        assert_eq!(stdout_receiver.recv().unwrap(), 99);
        assert_eq!(program.memory, vec![4, 2, 99]);
    }

    #[test]
    fn test_op5_jump_if_true_1() {
        let mut program = create_program(vec![1105, 1, 4, -1, 99]);
        program.run();
    }

    #[test]
    fn test_op5_jump_if_true_2() {
        let mut program = create_program(vec![1105, 0, 4, 99, -1]);
        program.run();
    }

    #[test]
    fn test_op5_jump_if_true_3() {
        let mut program = create_program(vec![1105, -1, 4, -1, 99]);
        program.run();
    }

    #[test]
    fn test_op6_jump_if_false_1() {
        let mut program = create_program(vec![1106, 0, 4, -1, 99]);
        program.run();
    }

    #[test]
    fn test_op6_jump_if_false_2() {
        let mut program = create_program(vec![1106, 1, 4, 99, -1]);
        program.run();
    }

    #[test]
    fn test_op6_jump_if_false_3() {
        let mut program = create_program(vec![1106, -1, 4, 99, -1]);
        program.run();
    }

    #[test]
    fn test_op7_less_than_1() {
        let mut program = create_program(vec![1107, -1, 3, 5, 99, -1]);
        program.run();
        assert_eq!(program.memory, vec![1107, -1, 3, 5, 99, 1]);
    }

    #[test]
    fn test_op7_less_than_2() {
        let mut program = create_program(vec![1107, 1, -3, 5, 99, -1]);
        program.run();
        assert_eq!(program.memory, vec![1107, 1, -3, 5, 99, 0]);
    }

    #[test]
    fn test_op8_equals_1() {
        let mut program = create_program(vec![1108, 3, 3, 5, 99, -1]);
        program.run();
        assert_eq!(program.memory, vec![1108, 3, 3, 5, 99, 1]);
    }

    #[test]
    fn test_op8_equals_2() {
        let mut program = create_program(vec![1108, 2, 3, 5, 99, -1]);
        program.run();
        assert_eq!(program.memory, vec![1108, 2, 3, 5, 99, 0]);
    }

    #[test]
    fn test_op9_relative_base_offset() {
        let mut program = create_program(vec![109, 2, 99]);
        program.run();
        assert_eq!(program.relative_base, 2);
    }

    #[test]
    fn test_op99_halt() {
        let mut program = create_program(vec![99, -1]);
        program.run();
    }

    #[test]
    fn test_input_output_program1() {
        // Test program from day 5
        let mem = vec![3, 9, 8, 9, 10, 9, 4, 9, 99, -1, 8];
        let (stdin_sender, stdin_receiver) = channel();
        let (stdout_sender, stdout_receiver) = channel();
        {
            let mut program = Program::new(mem.clone(), stdin_receiver, stdout_sender);
            stdin_sender.send(8).unwrap();
            program.run();
        }
        assert_eq!(stdout_receiver.into_iter().collect::<Vec<_>>(), vec![1]);
    }

    // #[test]
    // fn test_
    // 3,9,7,9,10,9,4,9,99,-1,8
}
